### PR TITLE
Ap is

### PR DESCRIPTION
**Pull request by gloriaodipo**

### CODE REVIEW

File: app/models.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+class Menu():
+    meals = []
+    def __init__(self):
+         meals = []
+    def add(self, meal):
+        meals.append(meal)
+    def remove(self, meal_id):
+        for meal in meals:
+            if meal_id == meal.meal_id:
+                del(meals[meal])
```
Comment (sirghiny) [inline app/models.py]: The deletion will rely on the index of the meal so it should be something like `del(meals[meals.index(meal)])`

File: app/models.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+#returns a list of all meals 
+    def get(self):
+        return self.meals
+
+    def edit(self,meal_id, meal_name, price, category):
```
Comment (sirghiny) [inline app/models.py]: Consider the possibility of editing only one field. This implementation will force the user to always update all fields. A hint would be to pass in a dictionary of the updated fields and only update their values.

File: app/models.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+class Order():
+    meals = []
+    def __init__(self, order_id):
+        self.order_id = order_id
+    def add(self,meal):
+        meals.append(meal)
+    def remove(self, meal):
+        for meal in meals:
+            if  meal_id == meals.meal_id:
+                del(meals[meal])
```
Comment (sirghiny) [inline app/models.py]: Also, the deletion will depend on an index, not the object itself..

File: app/views.py [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    def get(self):
+        meal = Meal_Schema(many = True)
+        meal_items = meal.dump(meals)
+
+        result = jsonify(meal_items.data)
+        result.status_code = 200
+        return result
+
+
+class Single_meal_API(Resource):
```
Comment (sirghiny) [inline app/views.py]: For all class names, you can use CamelCase e.g in this case, let it be SingleMealResource. Only use snake_case in methods and functions.

File: app/views.py [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+                result.status_code = 200
+                return result 
+            else:
+                result = jsonify({"message": "invalid selection"})
+                result.status_code = 404
+                return result 
+
+    
+        
+api.add_resource(User_signup_API, '/user/api/v1/signup')
```
Comment (sirghiny) [inline app/views.py]: Be consistent in your routes. They should all start with the version e.g `/api/v1/user/signup`

File: app/views.py [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+app = Flask(__name__)
+api = Api(app)
+
+users = [User(1, "gloria", "odipo", "gloriaodipo", "godipo@gmail.com", "bubble")]
+meals = [Meal(2, "beef with rice", 500.00, "main dish")]
+orders = []
+
+class User_signup_API(Resource):
+    def post(self):
+        user = json.loads(request.data)
```
Comment (sirghiny) [inline app/views.py]: Add a condition for ensuring all information is provided. Upon running it, it registered without data.

File: app/views.py [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+
+        users.append(u)
+
+        result=jsonify ({"message": "successfully registered"})
+        result.status_code = 201
+        return result
+
+class User_login_API(Resource):
+    def post(self, user_id):
+        access = json.loads(request.data)
```
Comment (sirghiny) [inline app/views.py]: For efficiency. just use request.get_json()

File: app/views.py [hunk 8]
```diff
@@ -2,50 +2,118 @@
...
+meals = [Meal(2, "beef with rice", 500.00, "main dish")]
+orders = []
+
+class User_signup_API(Resource):
     def post(self):
-        pass
-   
-class MealsAPI(Resource):
+        user = json.loads(request.data)
+        u = User(user_id=user.get('user_id'), first_name=user.get('first_name'), last_name=user.get('last_name'),
```
Comment (sirghiny) [inline app/views.py]: Make `user_id` be generated automatically e.g by using the length of the user's list.

File: app/views.py [hunk 9]
```diff
@@ -2,50 +2,118 @@
...
     def post(self):
-        pass
-
-    def get(self):
-        pass
-
-    def get(self, id):
-        pass
+        meal = json.loads(request.data)
+        m = Meal(meal_id=meal.get('meal_id'), meal_name=meal.get('meal_name'), price=meal.get('price'),
```
Comment (sirghiny) [inline app/views.py]: Make `mealr_id` be generated automatically as suggested in the `user_id` section.

File: app/views.py [hunk 10]
```diff
@@ -2,50 +2,118 @@
...
+                result.status_code = 200
+                return result 
+            else:
+                result = jsonify({"message": "invalid selection"})
+                result.status_code = 404
+                return result 
 
     
         
+api.add_resource(User_signup_API, '/user/api/v1/signup')
```
Comment (sirghiny) [inline app/views.py]: Be consistent with the routes, e.g this should be `/api/v1/user/signup/`

**Comment by sirghiny** [comment]
Generally, address styling issues. You can use `pylint` for in-depth analysis or `flake8` for the simpler styling.

File: tests.py [hunk 11]
```diff
@@ -19,22 +24,22 @@ def setUp(self):
     
     def test_login(self):
         """Test API can successfully log in registered users using username and password (POST request)"""
-        response = self.app.post('/api/v1/login')
+        response = self.client.post('/api/v1/user/login/1')
```
Comment (sirghiny) [inline tests.py]: Present the data being sent here e.g `self.client.post('/api/v1/user/login/1', data=self.data, content_type='application/json')`

File: tests.py [hunk 12]
```diff
@@ -19,22 +24,22 @@ def setUp(self):
...
         result = json.loads(response.data)
         self.assertEqual(result["message"], "You are successfully logged in")
         self.assertEqual(response.status_code, 200)
 
     def test_wrong_login(self):
         """Test API cannot authenticate login when wrong password is used or no password supplied (POST request)"""
-        response = self.app.post('/api/v1/login')
+        response = self.client.post('/api/v1/user/login/4')
```
Comment (sirghiny) [inline tests.py]: Pass data here too.

File: app/views.py [hunk 13]
```diff
@@ -1,48 +1,158 @@
...
     def post(self):
-        pass
-   
+        user = request.get_json()
+        u = User(first_name=user.get('first_name'), last_name=user.get('last_name'),
+        username=user.get('username'), email=user.get('email'), password=user.get('password'))
+
+        users.append(u)
+
+        result=jsonify ({"message": "successfully registered"})
```
Comment (sirghiny) [inline app/views.py]: Ensure the returned json  equals the one in the tests self.data. This only returns the message whereas the test payload has user information.

File: tests.py [hunk 14]
```diff
@@ -110,7 +129,7 @@ def test_delete_meal(self):
 
     def test_delete_invalid_meal(self):
         """Test API can return a 204:no content when deleting a meal that's inexistent"""
-        response = self.app.delete('/api/v1/meals/<int:2>', data = json.dumps(self.data) , content_type = 'application/json')
+        response = self.client.delete('/api/v1/meals/7', data = json.dumps(self.data) , content_type = 'application/json')
```
Comment (sirghiny) [inline tests.py]: `DELETE` does not need any data, just the `meal_id`, this will cause it to expect a nonexistent `json`

File: tests.py [hunk 15]
```diff
@@ -100,7 +119,7 @@ def test_update_meal(self):
 
     def test_delete_meal(self):
         """Test API can delete a meal using meal_id (DELETE request)"""
-        response = self.app.delete('/api/v1/meals/<int:1>', data = json.dumps(self.data) , content_type = 'application/json')
+        response = self.client.delete('/api/v1/meals/2', data = json.dumps(self.data) , content_type = 'application/json')
```
Comment (sirghiny) [inline tests.py]: `DELETE` needs no `data`, just the `meal_id`

File: tests.py [hunk 16]
```diff
@@ -90,7 +109,7 @@ def test_get_one_meal(self):
 
     def test_update_meal(self):
         """Test API can modify/update details of a given meal using meal_id (PUT request)"""
-        response = self.app.put('/api/v1/meals/<int:1>', data = json.dumps(self.data) , content_type = 'application/json')
+        response = self.client.put('/api/v1/meals/2', data = json.dumps(self.data) , content_type = 'application/json')
```
Comment (sirghiny) [inline tests.py]: Ensure the data passed in a put request is different from the data used to create the object to notice a change.

File: app/views.py (hunk 13 above)
Comment (gloriaodipo) [reply app/views.py]: I get this error when i run my tests: ValueError: No JSON object could be decoded. how do i ensure the returned json equals the one in the test's self.data

**Pull request merged**